### ISSUE TITLE
Aztec C compiler doesn't work

### ISSUE THREAD

**Issue by skx** [skx/cpmulator#38]
The assembler and compiler _appear_ to produce identical output.

However the link step fails.

Sample files here:

* https://github.com/skx/z80-playground-cpm-fat/tree/main/dist/CPM/DISKS/C

Try running these in the CCP and initial state looks good, but then:

```
$ ./cpmulator -directories -cd ~/Repos/github.com/skx/z80-playground-cpm-fat/dist/CPM/DISKS/

A>c:

C>CC ECHO

C Vers. 1.06D 8080  (C) 1982 1983 1984 by Manx Software Systems

A>AS ECHO

AS?

A>C:

C>AS ECHO

8080 Assembler Vers. 1.06D

A>C:

C>LN ECHO.O T.LIB C.LIB

C Linker Vers. 1.06D
Base: 0100   Code: 008b  Data: 0000  Udata: 0000  Total: 00008e

A>C:

C>ECHO HELLO WORLD


C>

```

**Comment by skx** [comment]
To compile a .C file into a .COM file there are three steps:

* Converter, takes .C and produces .ASM
* Compiler,  takes .ASM and produces .O
* Linker.  Does magic.

Using the `hello.C` file I see that the generated output of my real hardware and this emulator is 100% identical.

However when compiling the .ASM file into a .O file we're missing contents.  The real hardware produces this:

```

00000000: 041a 0000 0000 0000 0000 2000 0000 0000  .......... .....
00000010: b100 4100 bd00 0001 1011 6000 10cd 7300  ..A.......`...s.
00000020: 1021 2161 0030 11e5 cd77 001f d121 0000  .!!a.0...w...!..
00000030: c948 656c 6c6f 2c20 776f 726c 1264 0a00  .Hello, worl.d..
00000040: 0007 0800 0000 002e 6265 6769 6e00 0007  ........begin...
00000050: 0800 0000 002e 6368 6c00 0000 0007 0800  ......chl.......
00000060: 0000 002e 7377 7400 0000 0007 0800 0000  ....swt.........
00000070: 0063 7361 7665 0000 0007 0800 0000 0063  .csave.........c
00000080: 7265 7400 0000 0007 0800 0000 002e 6d6f  ret...........mo
00000090: 7665 0000 0002 8800 0000 006d 6169 6e5f  ve.........main_
000000a0: 0000 0007 0800 0000 0070 7269 6e74 665f  .........printf_
000000b0: 0001 4000 0000 0002 c012 0000 0000 0000  ..@.............
000000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000100: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000120: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000140: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000150: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000160: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

My emulator produces this:

```
00000000: 041a 0000 0000 0000 0000 2000 0000 0000  .......... .....
00000010: b100 4100 bd00 0001 1011 6000 10cd 7300  ..A.......`...s.
00000020: 1021 2161 0030 11e5 cd77 001f d121 0000  .!!a.0...w...!..
00000030: c948 656c 6c6f 2c20 776f 726c 1264 0a00  .Hello, worl.d..
00000040: 0007 0800 0000 002e 6265 6769 6e00 0007  ........begin...
00000050: 0800 0000 002e 6368 6c00 0000 0007 0800  ......chl.......
00000060: 0000 002e 7377 7400 0000 0007 0800 0000  ....swt.........
00000070: 0063 7361 7665 0000 0007 0800 0000 0063  .csave.........c
00000080: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000090: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
000000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000100: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000120: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000140: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000150: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000160: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

NOTE: The real compiler has references to "main_" and "printf_" as you can see.  But my output does not, and that leads the the linker error we see:

```
C>ln hello.o t.lib c.lib

C Linker Vers. 1.06D
Undefined symbol: c
Undefined symbol: main_
```

TLDR:

* CC.com -> works
* AS.com -> produces malformed .O file
* LN.com -> works, if I feed it a correct .O file, produced elsewhere.

So the issue here is that the assembler is not working,
```

```

**Comment by skx** [comment]
The more I look into this the more confused I get, but I did spot two problems with my current code:

* Syscall 11 I've implemented as `SETDMA`, but it _should_ be `C_STAT`.
  * I guess I looked at bios not bdos.  Oops.
* Syscall 13, DRV_ALLRESET should reset the DMA address to the default of 0x80 in addition to changing the default disk and usernumber to zero.
* I wonder if I need to care about syscall 24 DRV_LOGINVEC, which I think is not doing the right thing - but nor is syscall 31 DRV_DPB

TLDR:

* Some syscalls are wrong.  Fix them.
* Get an output of what offsets/blocks I'm reading and writing, along with their contents.
* Get the same output for the real (hardware) system I use, and compare.

I suspect it it something odd with the random-write, after all the file is created in the right way, but the contents of the second & third blocks are wrong:

* Either I've messed up the DMA address, or I've messed up the offset calculation in some way.

That said _reading_ seems OK.

**Comment by skx** [comment]
So the previous cleanup didn't help me out in terms of functionality.

So I've confirmed that the C compiler has three steps:

1.  CC.com 
   * Reads HELLO.C, creates HELLO.ASM
2. AS.com
   * Reads HELLO.ASM creates HELLO.O
3. LN.COM
   * Reads HELLO.O produces HELLO.COM

Steps 1 & 3 are fine.  If I replace the "HELLO.O" I generate with one produced on real hardware all is fine.  The resulting binary is OK.

The problem is obviously the assembler is producing bogus output, but I still can't see why.  Tracing through there is a lot of functions called, but I guess I need to add diagnostics to more of them - the DMA setup now seems correct and matches real hardware at least.
